src/lsm/forest.zig

```diff
@@ -162,22 +179,506 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+
+                // FIXME: This wastes the remainder.
+                const target_value_blocks = blocks[2 + equal_split_count + equal_split_count ..][0..equal_split_count];
+
+                std.log.info("divide_blocks: block_count_bar={} source_index_blocks.len={} source_value_level_a.len={} source_value_level_b.len={} target_value_blocks.len={}", .{
+                    block_count_bar,
+                    source_index_blocks.len,
+                    source_value_level_a.len,
+                    source_value_level_b.len,
+                    target_value_blocks.len,
+                });
+
+                return .{
+                    .source_index_blocks = source_index_blocks,
+                    .source_value_blocks = .{
+                        CompactionHelper.BlockFIFO.init(source_value_level_a),
+                        CompactionHelper.BlockFIFO.init(source_value_level_b),
+                    },
+                    .target_value_blocks = CompactionHelper.BlockFIFO.init(target_value_blocks),
+                };
+            }
+
+            pub fn beat(
+                self: *CompactionPipeline,
+                forest: *Forest,
+                op: u64,
+                callback: Callback,
+            ) void {
+                const compaction_beat = op % constants.lsm_batch_multiple;
+                const first_beat = compaction_beat == 0;
+                const half_beat = compaction_beat == @divExact(constants.lsm_batch_multiple, 2);
+
+                self.slot_filled_count = 0;
+                self.slot_running_count = 0;
+
+                if (first_beat or half_beat) {
+                    if (first_beat)
+                        self.bar_active = CompactionBitset.initEmpty();
+
+                    var compaction_blocks = self.compaction_blocks;
+                    for (self.compactions.slice(), 0..) |*compaction, i| {
+                        if (compaction.info.level_b % 2 == 0 and first_beat) continue;
+                        if (compaction.info.level_b % 2 != 0 and half_beat) continue;
+
+                        // TODO: It's too easy to mess up the ranges, and give overlapping blocks. To be able to assert
+                        // that all our blocks are distinct would be great. Both here, and for the beat blocks set up
+                        // by divide_blocks...
+                        const immutable_table_a_block = compaction_blocks[compaction_blocks.len - 1].block;
+                        log.info("Immutable dummy block is: [{}]", .{compaction_blocks.len - 1});
+
+                        const blocks = compaction_blocks[compaction_blocks.len - 3 .. compaction_blocks.len - 1];
+                        log.info("Index block range is: [{}..{}]", .{ compaction_blocks.len - 3, compaction_blocks.len - 1 });
+                        for (blocks) |*block| block.next = null;
+
+                        compaction_blocks.len -= 3;
```

Comment:
Oh dear, didn't realize one can just assign to a slice length this way :fearful: 

Here be :dragon: :dragon:  indeed!